File: frontend/src/types/chain.ts
// Chain Types
// Mirrors: project/src/daemon/types/chain.rs

import type { TransactionViewResponse } from './transaction';

export interface ChainStatusResponse {
  block_count: number;
  is_valid: boolean;
  last_block_hash?: string;
  last_block_date?: string;
}

export interface BlockInfo {
  height: number;
  hash: string;
  prev_hash: string;
  merkle_root: string;
  nonce: number;
  timestamp: string;
  transactions: TransactionViewResponse[];
  size_bytes: number;
}

export interface ChainShowResponse {
  blocks: BlockInfo[];
}

export interface UtxoInfo {
  tx_id: string;
  index: number;
  value: number;
  address: string;
}

export interface UtxosParams {
  limit?: number;
}

export interface UtxosResponse {
  utxos: UtxoInfo[];
  total_value: number;
}

File: project/src/daemon/types/chain.rs
// Chain RPC Types
use serde::{Deserialize, Serialize};

use crate::daemon::types::TransactionViewResponse;

#[derive(Serialize, Deserialize, Debug)]
pub struct ChainStatusResponse {
    pub block_count: usize,
    pub is_valid: bool,
    pub last_block_hash: Option<String>,
    pub last_block_date: Option<String>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct BlockInfo {
    pub height: usize,
    pub hash: String,
    pub prev_hash: String,
    pub merkle_root: String,
    pub nonce: u32,
    pub timestamp: String,
    pub transactions: Vec<TransactionViewResponse>,
    pub size_bytes: usize,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct ChainShowResponse {
    pub blocks: Vec<BlockInfo>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct UtxoInfo {
    pub tx_id: String,
    pub index: usize,
    pub value: i64,
    pub address: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct UtxosParams {
    #[serde(default = "default_limit")]
    pub limit: u32,
}

fn default_limit() -> u32 {
    20
}

#[derive(Serialize, Deserialize, Debug)]
pub struct UtxosResponse {
    pub utxos: Vec<UtxoInfo>,
    pub total_value: i64,
}
